### PR TITLE
Modular build for ukvm

### PR DESCRIPTION
**Pull request by djwillia**
This PR should be considered with the matching one in djwillia/mirage#solo5. (https://github.com/djwillia/mirage/pull/5)

Most of the changes here are due to moving out the tests from the `kernel` directory, because each test will now have its own `ukvm-bin` generated.  `ukvm-configure` is the script that is called to generate the Makefile to build `ukvm-bin`.

### CODE REVIEW

File: ukvm/ukvm-configure [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    echo "    UKVM_SRC is /path/to/ukvm"
+    echo "    MODULES can be any combination of: net blk gdb"
+    exit
+fi
+
+UKVM_SRC=`readlink -e $1`
+shift
+UKVM_MODULES=$@
+
+(
```
Comment (ricarkol) [inline ukvm/ukvm-configure]: You can simplify this a lot by using make string operations, and possibly not even needing to have a generated Makefile.ukvm. Something like this:

`UKVM_MODULES=blk net
all:
        gcc $(addprefix -DUKVM_MODULE_,${UKVM_MODULES}) $(addprefix _build-ukvm/ukvm-,${UKVM_MODULES})`

File: ukvm/ukvm-core.c [hunk 2]
```diff
@@ -479,7 +479,7 @@ static void setup_cpuid(int kvm, int vcpufd)
         err(1, "KVM_SET_CPUID2 failed");
 }
 
-
+#if 0
```
Comment (ricarkol) [inline ukvm/ukvm-core.c]: I think we will never ever use this functionality. You might as well delete it. If we want it back we have the commits history.

File: Makefile [hunk 3]
```diff
@@ -104,7 +107,10 @@ opam-ukvm-install: solo5-kernel-ukvm.pc ukvm_target
 	cp ukvm/ukvm.h $(OPAM_UKVM_INCDIR)/ukvm.h
 	cp kernel/ukvm/solo5.o kernel/ukvm/solo5.lds $(OPAM_UKVM_LIBDIR)
 	mkdir -p $(OPAM_BINDIR)
-	cp ukvm/ukvm $(OPAM_BINDIR)
+#	cp ukvm/ukvm $(OPAM_BINDIR)
```
Comment (mato) [inline Makefile]: Remove rather than comment out?

File: README.md [hunk 4]
```diff
@@ -97,9 +94,14 @@ use to talk to the unikernel:
...
 
+A specialized monitor called `ukvm-bin` is generated as part of the
+build process, so the command line arguments may differ depending on
+the needs of the unikernel.  To see the arguments, run `ukvm-bin` with
+no arguments or `--help`.
+
 To launch the unikernel:
 
-    ukvm /dev/zero tap100 ./mir-stackv4.ukvm
+    ukvm-bin --net=tap100 ./mir-stackv4.ukvm
```
Comment (mato) [inline README.md]: Should be `./ukvm-bin` (unless you have `.` in `$PATH` which is generally considered a bad idea).

File: README.md [hunk 5]
```diff
@@ -138,12 +140,35 @@ Refer to the
...
+You can debug the unikernel running in ukvm using gdb, but need to
+build a `ukvm-bin` with gdb support.  To do so, edit the Makefile
+(generated in the case of Mirage), and add `gdb` to the `UKVM_MODULES`
+make variable.  Then, running `make ukvm-bin` should build a version
+with gdb support.
 
-    sudo $(which ukvm) ~/disk.img tap100 mir-console.solo5-ukvm --gdb
+Start ukvm with the `--gdb` flag, like this:
+
+    ukvm-bin mir-console.solo5-ukvm --gdb
```
Comment (mato) [inline README.md]: `./ukvm-bin`

**Comment by mato** [comment]
@djwillia, this looks good to me, good stuff! I've tested with the companion PR (djwillia/mirage#5) and everything works as expected.

One important piece that is missing is that we can no longer build only `virtio` or `ukvm` and the corresponding tests. This is important at least for @hannesm FreeBSD build. I've made some followup fixes for this and other cleanups to the Makefiles, at https://github.com/mato/solo5/commit/9ce1e32842552d1d007acc7278917815d7dff4f3.

In the interests of getting this into an initial OPAM release of Mirage/Solo5 (#36), I think we should merge this and the companion djwillia/mirage#5 as is, followed up by a PR for my changes above, then I can cut an initial release and submit `solo5-kernel-{ukvm,virtio}` to opam.ocaml.org.

Dan, please confirm is this is OK with you?

File: ukvm/ukvm-core.c (hunk 2 above)
Comment (djwillia) [inline ukvm/ukvm-core.c]: @ricarkol do you mean the interrupt inject code?  perhaps that should be a different issue/PR?

File: ukvm/ukvm-configure (hunk 1 above)
Comment (djwillia) [inline ukvm/ukvm-configure]: @ricarkol @mato If I understand what you are saying @ricarkol, the `Makefile.ukvm` would be the same in all cases.  The only thing that would change is what `UKVM_MODULES` was set to before the `include Makefile.ukvm`.  Is that right?

File: ukvm/ukvm-core.c (hunk 2 above)
Comment (ricarkol) [inline ukvm/ukvm-core.c]: Yeap, agree: let's have a another one just for that instead (cleanup), which doesn't have to be now.

File: ukvm/ukvm-configure (hunk 1 above)
Comment (ricarkol) [inline ukvm/ukvm-configure]: Yes, the suggestion is to have one Makefile.ukvm in all cases.

Ah, if you decide to use the example, you will have to change the names to UKVM_OBJECTS (and make the MODULES upper-case). Thanks @mato

**Comment by djwillia** [comment]
Thanks for the comments, it's OK with me to go ahead and merge, put your Makefile changes in and cut the initial release!

**Comment by mato** [comment]
@djwillia Thanks for confirming, merged!

**Pull request merged**